bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java

```diff
@@ -139,12 +139,18 @@ public void flush() {
             ByteBuffer buffer = outboundBuffer.lockAndGet();
             if (buffer == null) {
                 // buffer has been released
+                outboundBuffer.unlock();
                 return;
             }
             buffer.flip();
             try {
                 while (buffer.hasRemaining()) {
-                    channel.write(buffer);
+                    int written = channel.write(buffer);
+                    // for unknown reason, -2 is returned of every call to channel.write, and no data is written, buffer.hasRemaining() is always true.
```

Comment:
`SocketChannel.write` returns `-2`??

Reply:
not sure about -2, but I also have an issue here, sometimes connection just stuck in infinite loop trying to write some messages, I've tried to add `key.interestOps(SelectionKey.OP_WRITE)`, so for now I fixed it for myself with just closing the connection as soon as I get `written == 0`, at least it doesn't consume cpu in the loop

Reply:
I tried to reproduce this problem, but failed.    
But this -2 infinite loop did happen in my previous tests, and this changes fixed it.    
Any way, negative returned value is always bad, and we should handle this.(Throw exception looks like a reasonable operation)

Some hints about negative values: 
In this file sun.nio.ch.IOStatus, some constants is defined:
![image](https://user-images.githubusercontent.com/6279792/114350109-58623780-9b9b-11eb-99ba-1c29b5df777c.png)

Reply:
what jvm do you use? In openjdk returned status is always normalized through 
```
//sun.nio.ch.IOStatus
    public static int normalize(int var0) {
        return var0 == -2 ? 0 : var0;
    }
```
so, actually -2 means 0, and it's ok for non-blocking sockets, output buffer is full and we should wait. But I still had the situation when OP_WRITE didn't happen for the socket

Reply:
here is the jdk version
java version "1.8.0_60"
Java(TM) SE Runtime Environment (build 1.8.0_60-b27)
Java HotSpot(TM) 64-Bit Server VM (build 25.60-b23, mixed mode)

Reply:
It's a mistake.
Maybe is observed a -2 is returned before normalize,but incorrectly recognize this value as returned by write.